### PR TITLE
Basic Weariness Implementation

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
SUMMARY: Features "Add weariness, a sort of long term stamina which limits the character's ability to be extremely active 24/7"

#### Purpose of change
Work towards #42343

#### Describe the solution

abe5051fca50403aacd33e29ce66e28d99abd906
Encapsulate and rename Character activity level
    
And some other minor tweaks.
    
In preparation for weariness changes, get some churn out of the way - rename activity_level to attempted_activity level, and add an accessor function - this is no different from accessing it directly at the moment, but that will change in the future.
    
Split character BMR calculation into two steps, base_bmr and get_bmr, which additionally adds enchantments and activity level to the BMR.
    
Stop resetting the activity level when activities end. It doesn't really make sense to do this, we don't reset the activity level in any other cases. It also is problematic for future weariness changes.
    
Clear sleep deprivation when clearing the player in tests too. No tests suffer from it now, but sleep deprivation not being cleared could cause characters to randomly fall asleep in tests...
    
Move the exertion increases from melee attacks from choosing to perform a melee attack to the melee attack function. This makes it more clear where this is happening, and makes it easier to tweak in the future.

cb1a54cffb730c34173d3f4c739e88eb5521c0aa
Minor test BMR adjustments
    
I've made some changes that change where the BMR is multiplied by the activity level, which means some minor floating point result changes.
    
Adjust the tests for these.

af03b1ef54bcf94e14fc9ec13c80a79652f9f9a7
Basic implementation of weariness
    
Add a struct weary to Character, containing information how many calories they have consumed and how many calories they have gained, in addition to timing info used to recover or otherwise modify this info.
    
Add some functions to character:
 - weariness_level(), reporting the weariness level of the character.
 - exertion_adjusted_move_multiplier(), reporting as a percentage (e.g. 0.01 = 1%) how fast an activity at a certain activity level will be due to weariness.
 - try_reduce_weariness(), which is called every 5 minutes and tries to reduce the weariness, or various things feeding into it.
 - maximum_exertion_level(), which returns the maximum exertion level allowed by how weary the character is.
 - debug_weary_info, which is to be used in tests.
 - suffer_from_exertion(), which adds effects and provides messaging for the effects of exertion.
    
Add effects to display weariness. I'm not particularly happen wih the names, descriptions, or apply messages, but it's better than nothing. the ones above level 5 will probably get some effects at some point.
    
Add two additional private functions that return the threshold for going to a new level of weariness, and the raw weariness.
    
Add some external options controlling the rate of gain (because screw recompling every time I want to change a number) (todo: cache these).
    
Based on how weary the character is, scale how much progress they make each turn during activities.

635c7a4c7493b88af3cda2b0939df8ab8b1b73ec
Add weariness tests
    
Add some tests outlining scenarios, and when a character transitions or how much weariness they gain from these.
    
Along with these, add a general infrastructure for performing activities and monitoring weariness during these activities.
    
Current values for tesks are based on what looks good - there isn't a strong data driven attachment to them.

3f6010a932f09dc7a6605cca24fc1261a210144b
Add transition_minutes to use Approx for comparing
    
Rather than do the approximation margin calculation inside is_transition_at, this alternative adds a transition_minutes function that returns the number of minutes where the given weariness transition first occurs in the weariness_events. This allows the CHECKs to use builtin Catch2 features for Approx() and margin().

e3cb90b0cf05b2daca600c6e1c0e865d8de8c776
Adjust weariness_events::transition_minutes
    
Make it instead return the time closest to the time passed to it, instead of the returned time.
    
I'm not super happy about duplicating the expected time each time it's checked, but I think this is a good way for it to return a value that is not the first transition.

f30af904829c20454dcbeb09fd6b4cf01d1695e2
Move weary activity handling code to separate file
    
This isn't really weariness specific, and might be more useful as general infrastructure. Move it from weariness specific files to a helper file and header pair.

5e0ad81085f8f6dd22184332db9cfa1e7a6fd1be
Convert schedule into a virtual class w/ overloads
    
To enable doing more than just activities, convert schedule to a virtual class, and call functions on them to perform their function.
    
Introduce a sleeping, eating, and clearing guts version, in addition to the activity version.
The clearing guts function is because clear_avatar() feeds the avatar, which may not be wanted.
    
Convert tasklist to use a schedule * to preserve the virtualized members.
    
Move all the schedules to static constants at the top of the file instead of creating them in each test case, for cleanliness and also because creating them in the function call no longer works when using pointers.

8e348997a7499ad2870978bed91e3ce5fa6d8a53
Move function definitions out of header
    
There's no real reason for these to be in the header instead of a C++ file.

dce9a33fc5d56bc147ab0d978206ddfc16c76ea8
Add vehicle modification weariness recovery test
    
Some minor infrastructure to:
Give the character a light and bag, so that activities that need light are not cancelled.
Move all the various schedules out of the TEST_CASEs and into static declarations at the top of the file.
    
Tweak the CHECKs on the digging_24h test to make sure it keeps passing.
    
Add a mechanic_day case, checking that having some decent meals and having a somewhat intensive day has a sane weariness recovery.

#### Testing
See tests.

#### Additional context
I'd like to add more tests, but I think our current coverage is good enough.

### CODE REVIEW

**Comment by Aivean** [comment]
That's a wonderful addition!

Just to be clear, what real-world phenomenon are you simulating? Muscle fatigue, neural fatigue or something that encompasses both?

**Comment by anothersimulacrum** [comment]
Whoops. I don't have a specific phenomena in mind, but this is closest to a model of muscle exhaustion.

**Comment by Aivean** [comment]
> Whoops. I don't have a specific phenomena in mind, but this is closest to a model of muscle exhaustion.

Why I'm asking, if you have something from the real world as a reference model, it should be easier to find reference numbers for the unit tests. There is plenty of sport data on the internet.

**Comment by anothersimulacrum** [comment]
I think this is ready.

**Comment by I-am-Erk** [comment]
@aivean It's an amalgam model for physical and mental fatigue in general. Overall it's hard to find good studies on things like work speed over the course of a day, but I did get a fair bit of data from studies on military entrenching. In the end, this is balanced less a perfect model of human performance and more as a gestalt, balanced in how it affects calorie expenditure.

**Comment by anothersimulacrum** [comment]
Seems to be an issue generating the display for the activity which leads to a crash sometimes.

**Comment by anothersimulacrum** [comment]
Well nevermind, I can't seem to reproduce that any longer.

**Comment by kevingranade** [comment]
Not necessary for this PR, but looking at the places that are annotated for weariness tracking, I think what we want to do eventually is disallow access to the moves member and hide it behind a method that you call to log that activity happened, like spend_moves( nominal_moves, activity_level ), that can make the adjustments for weariness, fatigue, whatever in a single place.

**Comment by anothersimulacrum** [comment]
Thanks for the merge!

File: tests/weary_test.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+static const activity_schedule task_wait( activity_id( "ACT_WAIT" ), 5_minutes );
+static const activity_schedule task_firstaid( activity_id( "ACT_FIRSTAID" ), 5_minutes );
+static const activity_schedule task_plant( activity_id( "ACT_PLANT_SEED" ), 5_minutes );
+static const activity_schedule task_weld( activity_id( "ACT_VEHICLE" ), 5_minutes );
+static const activity_schedule task_read( activity_id( "ACT_READ" ), 5_minutes );
+
+static const meal_schedule sausage( itype_id( "sausage" ) );
+static const meal_schedule milk( itype_id( "milk" ) );
+
+static const sleep_schedule sleep{};
```
Comment (ZhilkinSerg) [inline tests/weary_test.cpp:22]: Breaks OSX builds as it already has `sleep` defined elsewhere:

```
06:51:38 ../tests/weary_test.cpp:22:29: error: redefinition of 'sleep' as different kind of symbol
06:51:38 static const sleep_schedule sleep{};
06:51:38                             ^
06:51:38 /usr/local/osxcross/osxcross/target/bin/../SDK/MacOSX10.11.sdk/usr/include/unistd.h:480:3: note: previous definition is here
06:51:38          sleep(unsigned int) __DARWIN_ALIAS_C(sleep);
06:51:38          ^
06:51:38 ../tests/weary_test.cpp:113:30: error: reference to type 'const schedule' could not bind to an lvalue of type 'unsigned int (unsigned int)'
06:51:38     mechanic_day.enschedule( sleep, 10_hours );
06:51:38                              ^~~~~
06:51:38 ../tests/activity_scheduling_helper.h:83:42: note: passing argument to parameter 'added' here
06:51:38         void enschedule( const schedule &added, time_duration how_long );
06:51:38                                          ^
06:51:38 2 errors generated.
```

**Pull request merged**